icepack/statistics.py

```diff
@@ -0,0 +1,183 @@
+            "Step Acceptance Threshold": 0.05,
+            "Radius Shrinking Threshold": 0.05,
+            "Radius Growing Threshold": 0.9,
+            "Radius Shrinking Rate (Negative rho)": 0.0625,
+            "Radius Shrinking Rate (Positive rho)": 0.25,
+            "Sufficient Decrease Parameter": 1e-4,
+            "Safeguard Size": 1e2,
+        },
+    },
+    "Status Test": {
+        "Gradient Tolerance": 1e-4,
+        "Step Tolerance": 5e-3,
+        "Iteration Limit": 50,
+    },
+    "General": {
+        "Print Verbosity": 0,
+        "Secant": {},
+    },
+}
+
+
+class MaximumProbabilityEstimator:
+    def __init__(self, problem, solver_type="rol", **kwargs):
+        from firedrake_adjoint import Control, ReducedFunctional
+
+        self._problem = problem
+        if isinstance(problem.controls, (firedrake.Function, firedrake.Constant)):
+            self._controls = problem.controls.copy(deepcopy=True)
+        else:
+            self._controls = [field.copy(deepcopy=True) for field in problem.controls]
+
+        self._state = self.problem.simulation(self.controls)
+        self._likelihood = self.problem.likelihood(self.state)
+        self._prior = self.problem.prior(self.controls)
+        J = firedrake.assemble(self._likelihood) + firedrake.assemble(self._prior)
+
+        if isinstance(self.controls, (firedrake.Function, firedrake.Constant)):
+            reduced_objective = ReducedFunctional(J, Control(self.controls))
+        else:
+            controls = [Control(field) for field in self.controls]
+            reduced_objective = ReducedFunctional(J, controls)
+
+        if isinstance(solver_type, str) and solver_type.lower() == "rol":
+            problem_wrapper = pyadjoint.MinimizationProblem(reduced_objective)
+
+            if "rol_options" in kwargs.keys():
+                options = kwargs["rol_options"]
+            else:
+                options = deepcopy(_default_rol_options)
+
+                test = options["Status Test"]
+                test["Step Tolerance"] = kwargs.get("step_tolerance", 5e-3)
+                test["Gradient Tolerance"] = kwargs.get("gradient_tolerance", 1e-4)
+                test["Iteration Limit"] = kwargs.get("max_iterations", 50)
+
+                general = options["General"]
+                general["Print Verbosity"] = int(kwargs.get("verbose", False))
+
+                algorithm = kwargs.get("algorithm", "trust-region")
+                if algorithm == "bfgs":
+                    options["Step"] = {
+                        "Type": "Line Search",
+                        "Line Search": {
+                            "Descent Method": {"Type": "Quasi-Newton Step"}
+                        },
+                    }
+                    memory = kwargs.get("memory", 10)
+                    general["Secant"] = {
+                        "Type": "Limited-Memory BFGS",
+                        "Maximum Storage": memory,
+                    }
```

Comment:
Some comments about why you've chosen these options would be great (and perhaps a useful reminder to future you about how to avoid the horrors of ROL)